File: ch4/dfs.ml
open Scanf
open Printf

type graph = int list array (* Adjacency List representation *)

let range n = List.init n (fun x -> x)

(** Returns a list of DFS orderings of all connected components *)
let dfs (adj : graph) : int list list =
  let n = Array.length adj in
  let visited = Array.make n false in
  let rec visit order u =
    visited.(u) <- true;
    adj.(u) |> List.fold_left (fun order v ->
      if visited.(v) then order
      else visit order v
    ) (u :: order)
  in
  range n |> List.fold_left (fun components u ->
    if visited.(u) then components
    else (visit [] u |> List.rev) :: components
  ) [] |> List.rev

(* Returns an array of colors of the vertices *)
let floodfill (adj : graph) : int array =
  let n = Array.length adj in
  let label = Array.make n (-1) in
  let rec visit u color =
    label.(u) <- color;
    adj.(u) |> List.iter (fun v ->
      if label.(v) = -1 then visit v color
    )
  in
  let ncc = ref 0 in
  for u = 0 to n - 1 do
    if label.(u) = -1 then begin
      incr ncc;
      visit u !ncc
    end
  done;
  label

(** Return a topological ordering of the DAG *)
let topo_sort (adj : graph) : int list =
  let n = Array.length adj in
  let visited = Array.make n false in
  let rec visit order u =
    visited.(u) <- true;
    u ::
      (adj.(u) |> List.fold_left (fun order v ->
        if visited.(v) then order
        else visit order v
      ) order)
  in
  range n |> List.fold_left (fun acc u ->
    if visited.(u) then acc
    else visit acc u 
  ) []

type vertex_state = White | Gray | Black

let graph_check (adj : graph) =
  let n = Array.length adj in
  let color = Array.make n White in
  let parent = Array.make n (-1) in
  let bidirectionals = ref [] in
  let backs = ref [] in
  let forward_cross = ref [] in
  let rec visit u =
    color.(u) <- Gray;
    adj.(u) |> List.iter (fun v ->
      match color.(v) with
      | White -> parent.(v) <- u; visit v
      | Gray ->
        if v = parent.(u) then bidirectionals := (u, v) :: !bidirectionals
        else backs := (u, v) :: !backs
      | Black -> forward_cross := (u, v) :: !forward_cross
    );
    color.(u) <- Black
  in
  range n |> List.iter (fun u -> if color.(u) = White then visit u);
  !bidirectionals, !backs, !forward_cross

(** Returns a list of all articulation points and a list of all bridges of the
    simple graph.
    For multigraphs, store edge ids to the adjacency lists to check for
    bidirectional edges instead of using [parent] *)
let articulation_point_and_bridge (adj : graph) : int list * (int*int) list =
  let n = Array.length adj in
  let low = Array.make n 0 in
  let num = Array.make n (-1) in
  let parent = Array.make n (-1) in
  let counter = ref 0 in
  let rec visit u =
    low.(u) <- !counter;
    num.(u) <- !counter;
    incr counter;
    adj.(u) |> List.iter (fun v ->
      if num.(v) = -1 then begin
        parent.(v) <- u;
        visit v;
        low.(u) <- min low.(u) low.(v)
      end
      else if v <> parent.(u) then (** Use edge id for multigraphs *)
        low.(u) <- min low.(u) num.(v)
    )
  in
  range n |> List.iter (fun u ->
    if num.(u) = -1 then begin
      counter := 0;
      visit u
    end
  );
  let articulation_points = range n |> List.filter (fun u ->
    if num.(u) = 0 then (* special case for root *)
      adj.(u) |> List.filter (fun v -> parent.(v) = u) |> List.length > 1
    else adj.(u) |> List.exists (fun v -> parent.(v) = u && low.(v) >= num.(u))
  ) in
  let bridges = range n |> List.map (fun u ->
    adj.(u)
      |> List.filter (fun v -> parent.(v) = u && low.(v) > num.(u))
      |> List.map (fun v -> (u, v))
  ) |> List.concat in
  articulation_points, bridges

(** Returns a list of all strongly connected components of the directed graph *)
let tarjan_scc (adj : graph) : int list list =
  let n = Array.length adj in
  let low = Array.make n 0 in
  let num = Array.make n (-1) in
  let counter = ref 0 in
  let stack = ref [] in
  let on_stack = Array.make n false in
  let sccs = ref [] in
  let rec visit u =
    low.(u) <- !counter;
    num.(u) <- !counter;
    incr counter;
    stack := u :: !stack;
    on_stack.(u) <- true;
    adj.(u) |> List.iter (fun v ->
      if num.(v) = -1 then begin
        visit v;
        low.(u) <- min low.(u) low.(v)
      end
      else if on_stack.(v) then
        low.(u) <- min low.(u) num.(v)
    );
    if low.(u) = num.(u) then begin
      let rec pop_scc acc = function
        | hd :: tl ->
          if hd = u then hd :: acc, tl
          else pop_scc (hd :: acc) tl
        | [] -> assert false
      in
      let scc, stack' = pop_scc [] !stack in
      stack := stack';
      scc |> List.iter (fun v -> on_stack.(v) <- false);
      sccs := scc :: !sccs
    end
  in
  range n |> List.iter (fun u -> if num.(u) = -1 then visit u);
  !sccs

let print message =
  printf "==================================\n";
  printf "%s\n" message;
  printf "==================================\n"


let () =
  let ic = Scanning.open_in "01_in.txt" in
  let sc = object method nf = bscanf ic end in
  let n = sc#nf " %d " (fun x -> x) in
  let adj = Array.init n (fun _ ->
    sc#nf " %d " (fun degree ->
      List.init degree (fun _ -> sc#nf " %d %d " (fun v w -> v))) (* Ignoring weight *)
  ) in
  
  print "Standard DFS Demo (the input graph must be UNDIRECTED)";
  let ccs = dfs adj in
  ccs |> List.iteri (fun i component ->
    printf "CC %d:" (i + 1);
    component |> List.iter (printf " %d");
    printf "\n"
  );
  ccs |> List.length |> printf "There are %d connected components\n";

  print "Flood Fill Demo (the input graph must be UNDIRECTED)";
  floodfill adj |> Array.iteri (printf "Vertex %d has color %d\n");
  
  print "Topological Sort (the input graph must be DAG)";
  topo_sort adj |> List.iter (printf " %d"); printf "\n";

  print "Graph Edges Property Check";
  let bidirectionals, backs, forward_cross = graph_check adj in
  bidirectionals |> List.iter (fun (u, v) ->
    printf " Bidirectional (%d, %d) - (%d, %d)\n" u v v u);
  backs |> List.iter (fun (u, v) ->
    printf " Back Edge (%d, %d) (Cycle)\n" u v
  );
  forward_cross |> List.iter (fun (u, v) ->
    printf " Forward/Cross Edge (%d, %d)\n" u v
  );

  print "Articulation Points & Bridges (the input graph must be UNDIRECTED)";
  let aps, bridges = articulation_point_and_bridge adj in
  printf "Bridges:\n";
  bridges |> List.iter (fun (u, v) -> printf " Edge (%d, %d) is a bridge\n" u v);
  printf "Articulation Points:\n";
  aps |> List.iter (printf " Vertex %d\n");

  print "Strongly Connected Components (the input graph must be DIRECTED)";
  tarjan_scc adj |> List.iteri (fun i scc ->
    printf "SCC %d:" (i + 1);
    scc |> List.iter (printf " %d");
    printf "\n"
  )

File: ch4/traversal/dfs.cpp
// February 2019 note:
// This code uses new C++17 structured binding
// use this compiler setting "g++ -O2 -std=gnu++17 {cpp17file}"

#include <bits/stdc++.h>
using namespace std;

// In this chapter, we will frequently use these three data type shortcuts
// They may look cryptic but shortcuts are useful in competitive programming
typedef pair<int, int> ii;
typedef vector<ii> vii;
typedef vector<int> vi;

const int DFS_WHITE = -1; // normal DFS, do not change this with other values (other than 0), because we usually use memset with conjunction with DFS_WHITE
const int DFS_BLACK = 1;

vector<vii> AL;

void printThis(string message) {
  printf("==================================\n");
  printf("%s\n", message.c_str());
  printf("==================================\n");
}

vi dfs_num; // this variable has to be global, we cannot put it in recursion
int numCC;

void dfs(int u) { // DFS for normal usage: as graph traversal algorithm
  printf(" %d", u);                              // this vertex is visited
  dfs_num[u] = DFS_BLACK;                        // mark u as visited
  for (auto &[v, w] : AL[u])                     // (neighbor, weight) pair
    if (dfs_num[v] == DFS_WHITE)                 // to avoid cycle
      dfs(v);                                    // recursively visits v
} // for simple graph traversal, we ignore the weight

// note: this is not the version on implicit graph
void floodfill(int u, int color) {
  dfs_num[u] = color;                            // not just a generic DFS_BLACK
  for (auto &[v, w] : AL[u])
    if (dfs_num[v] == DFS_WHITE)
      floodfill(v, color);
}

vi topoSort; // global vector to store the toposort in reverse order

void dfs2(int u) { // change function name to differentiate with original dfs
  dfs_num[u] = DFS_BLACK;
  for (auto &[v, w] : AL[u])
    if (dfs_num[v] == DFS_WHITE)
      dfs2(v);
  topoSort.push_back(u);                         // this is the only change
}

#define DFS_GRAY 2                               // one more color
vi dfs_parent; // to differentiate real back edge versus bidirectional edge

void graphCheck(int u) { // DFS for checking graph edge properties
  dfs_num[u] = DFS_GRAY; // color this as DFS_GRAY (temp) instead of DFS_BLACK
  for (auto &[v, w] : AL[u]) {
    if (dfs_num[v] == DFS_WHITE) {               // Tree Edge
      dfs_parent[v] = u;                         // parent of v is me (u)
      graphCheck(v);
    }
    else if (dfs_num[v] == DFS_GRAY) {           // DFS_GRAY to DFS_GRAY
      if (v == dfs_parent[u])                    // to differentiate
        printf(" Bidirectional (%d, %d) - (%d, %d)\n", u, v, v, u);
      else  // the most frequent application: check if the given graph is cyclic
        printf(" Back Edge (%d, %d) (Cycle)\n", u, v);
    }
    else if (dfs_num[v] == DFS_BLACK)            // DFS_GRAY to DFS_BLACK
      printf(" Forward/Cross Edge (%d, %d)\n", u, v);
  }
  dfs_num[u] = DFS_BLACK; // after recursion, color this as DFS_BLACK (DONE)
}

vi dfs_low; // additional information for articulation points/bridges/SCCs
vi articulation_vertex;
int dfsNumberCounter, dfsRoot, rootChildren;

void articulationPointAndBridge(int u) {
  dfs_low[u] = dfs_num[u] = ++dfsNumberCounter;  // dfs_low[u] <= dfs_num[u]
  for (auto &[v, w] : AL[u]) {
    if (dfs_num[v] == DFS_WHITE) {               // a tree edge
      dfs_parent[v] = u;
      if (u == dfsRoot) ++rootChildren;          // special case, root

      articulationPointAndBridge(v);

      if (dfs_low[v] >= dfs_num[u])              // for articulation point
        articulation_vertex[u] = true;           // store this info first
      if (dfs_low[v] > dfs_num[u])               // for bridge
        printf(" Edge (%d, %d) is a bridge\n", u, v);
      dfs_low[u] = min(dfs_low[u], dfs_low[v]);  // update dfs_low[u]
    }
    else if (v != dfs_parent[u])                 // a back edge, not direct cycle
      dfs_low[u] = min(dfs_low[u], dfs_num[v]);  // update dfs_low[u]
  }
}

vi S, visited;                                   // additional global variables
int numSCC;

void tarjanSCC(int u) {
  dfs_low[u] = dfs_num[u] = ++dfsNumberCounter;  // dfs_low[u] <= dfs_num[u]
  S.push_back(u); // stores u in a vector based on order of visitation
  visited[u] = 1;
  for (auto &[v, w] : AL[u]) {
    if (dfs_num[v] == DFS_WHITE)
      tarjanSCC(v);
    if (visited[v])                              // condition for update
      dfs_low[u] = min(dfs_low[u], dfs_low[v]);
  }

  if (dfs_low[u] == dfs_num[u]) {                // a root/start of an SCC
    printf("SCC %d:", ++numSCC);                 // when recursion unwinds
    while (1) {
      int v = S.back(); S.pop_back(); visited[v] = 0;
      printf(" %d", v);
      if (u == v) break;
    }
    printf("\n");
  }
}

int main() {
  /*
  // Use the following input:
  // Graph in Figure 4.1
  9
  1 1 0
  3 0 0 2 0 3 0
  2 1 0 3 0
  3 1 0 2 0 4 0
  1 3 0
  0
  2 7 0 8 0
  1 6 0
  1 6 0

  // Example of directed acyclic graph in Figure 4.4 (for toposort)
  8
  2 1 0 2 0
  2 2 0 3 0
  2 3 0 5 0
  1 4 0
  0
  0
  0
  1 6 0

  // Example of directed graph with back edges
  3
  1 1 0
  1 2 0
  1 0 0

  // Left graph in Figure 4.6/4.7/4.8
  6
  1 1 0
  3 0 0 2 0 4 0
  1 1 0
  1 4 0
  3 1 0 3 0 5 0
  1 4 0

  // Right graph in Figure 4.6/4.7/4.8
  6
  1 1 0
  5 0 0 2 0 3 0 4 0 5 0
  1 1 0
  1 1 0
  2 1 0 5 0
  2 1 0 4 0

  // Directed graph in Figure 4.9
  8
  1 1 0
  1 3 0
  1 1 0
  2 2 0 4 0
  1 5 0
  1 7 0
  1 4 0
  1 6 0
  */

  freopen("dfs_in.txt", "r", stdin);

  int V; scanf("%d", &V);
  AL.assign(V, vii()); // assign blank vectors of pair<int, int>s to Al
  for (int u = 0; u < V; ++u) {
    int total_neighbors; scanf("%d", &total_neighbors);
    for (int j = 0; j < total_neighbors; ++j) {
      int v, w; scanf("%d %d", &v, &w);
      AL[u].emplace_back(v, w);
    }
  }

  printThis("Standard DFS Demo (the input graph must be UNDIRECTED)");
  numCC = 0;
  dfs_num.assign(V, DFS_WHITE);
  for (int u = 0; u < V; ++u)                    // for each u in [0..V-1]
    if (dfs_num[u] == DFS_WHITE)                 // if that u is unvisited
      printf("CC %d:", ++numCC), dfs(u), printf("\n"); // 3 lines here!
  printf("There are %d connected components\n", numCC);

  printThis("Flood Fill Demo (the input graph must be UNDIRECTED)");
  numCC = 0;
  dfs_num.assign(V, DFS_WHITE);
  for (int u = 0; u < V; ++u)
    if (dfs_num[u] == DFS_WHITE)
      floodfill(u, ++numCC);
  for (int u = 0; u < V; ++u)
    printf("Vertex %d has color %d\n", u, dfs_num[u]);

  // make sure that the given graph is DAG
  printThis("Topological Sort (the input graph must be DAG)");
  topoSort.clear();
  dfs_num.assign(V, DFS_WHITE);
  for (int u = 0; u < V; ++u)                    // same as finding CCs
    if (dfs_num[u] == DFS_WHITE)
      dfs2(u);
  reverse(topoSort.begin(), topoSort.end());     // reverse topoSort or
  for (auto &u : topoSort)                       // simply read the content
    printf(" %d", u);                            // of `topoSort' backwards
  printf("\n");

  printThis("Graph Edges Property Check");
  numCC = 0;
  dfs_num.assign(V, DFS_WHITE); dfs_parent.assign(V, -1);
  for (int u = 0; u < V; ++u)
    if (dfs_num[u] == DFS_WHITE)
      printf("Component %d:\n", ++numCC), graphCheck(u); // 2 lines in one

  printThis("Articulation Points & Bridges (the input graph must be UNDIRECTED)");
  dfsNumberCounter = 0; dfs_num.assign(V, DFS_WHITE); dfs_low.assign(V, 0);
  dfs_parent.assign(V, -1); articulation_vertex.assign(V, 0);
  printf("Bridges:\n");
  for (int u = 0; u < V; ++u)
    if (dfs_num[u] == DFS_WHITE) {
      dfsRoot = u; rootChildren = 0;
      articulationPointAndBridge(u);
      articulation_vertex[dfsRoot] = (rootChildren > 1); // special case
    }
  printf("Articulation Points:\n");
  for (int u = 0; u < V; ++u)
    if (articulation_vertex[u])
      printf(" Vertex %d\n", u);

  printThis("Strongly Connected Components (the input graph must be DIRECTED)");
  dfs_num.assign(V, DFS_WHITE); dfs_low.assign(V, 0); visited.assign(V, 0);
  dfsNumberCounter = numSCC = 0;
  for (int u = 0; u < V; ++u)
    if (dfs_num[u] == DFS_WHITE)
      tarjanSCC(u);

  return 0;
}
